gordon/main.py

```diff
@@ -83,20 +86,74 @@ def setup(config_root=''):
     return config
 
 
-def _log_or_exit_on_exceptions(errors, debug):
+def _log_or_exit_on_exceptions(base_msg, exc, debug):
     log_level_func = logging.warn
     if not debug:
         log_level_func = logging.error
 
-    base_msg = 'Plugin "{name}" was not loaded:'
-    for name, exc in errors:
-        msg = base_msg.format(name=name)
-        log_level_func(msg, exc_info=exc)
+    if isinstance(exc, list):
+        for exception in exc:
+            log_level_func(base_msg, exc_info=exception)
+    else:
+        log_level_func(base_msg, exc_info=exc)
 
     if not debug:
         raise SystemExit(1)
 
 
+def _assert_required_plugins(installed_plugins, debug):
+    # enricher not required
+    required_providers_available = {
+        'event_consumer': False,
+        'publisher': False,
+    }
+    for plugin in installed_plugins:
+        if interfaces.IEventConsumerClient.providedBy(plugin):
+            required_providers_available['event_consumer'] = True
+        elif interfaces.IPublisherClient.providedBy(plugin):
+            required_providers_available['publisher'] = True
+
+    missing = []
+    msg = ('The provider for the "{name}" interface is not configured for the '
+           'Gordon service or is not implemented.')
+    for provider, available in required_providers_available.items():
+        if not available:
+            exc = exceptions.MissingPluginError(msg.format(name=provider))
+            missing.append(exc)
+    if missing:
+        base_msg = 'Problem running plugins: '
+        _log_or_exit_on_exceptions(base_msg, missing, debug=debug)
+
+
+def _gather_runnable_plugins(plugins, debug):
+    plugins_to_run = []
+    for plugin in plugins:
+        if interfaces.IEventConsumerClient.providedBy(plugin):
+            # TODO (lynn): this should be switched out for adding
+            # the "verify interface implementation" ability. See
+            # https://docs.zope.org/zope.interface/verify.html
+            if not hasattr(plugin, 'run') or \
+                    not asyncio.iscoroutinefunction(plugin.run):
+                msg = (f'Implemention "{plugin}" of the required '
+                       '"IEventConsumerClient" interface does not have the '
+                       'necessary `run` method.')
+                exc = exceptions.InvalidPluginError(msg)
+                _log_or_exit_on_exceptions(msg, exc, debug)
+                continue
+            plugins_to_run.append(plugin)
+        elif hasattr(plugin, 'run') and asyncio.iscoroutinefunction(plugin.run):
```

Comment:
This was the spot I had in mind for the warning - if a plugin isn't implementing an interface, and has run(), but it's not async, it would be helpful to know that for debuggin.

Reply:
Hmm I get what you mean but I don't feel comfortable assuming that a synchronous `run` for an implementation wasn't on purpose.